Product Tags: implemented Networking, Yosemite and Storage layers

pmusolino opened the pull request:
Part of #2081 

## Description
First of all, I want to say sorry for all these changes (they are mainly mock JSON files and changes that comes from the edits to the Storage model). Initially, I thought that I needed to touch only the Networking and the Yosemite layers, but while I was working on it, I needed to touch also the Storage layers because I needed to change the relation between Products and Tags. In this PR, I implemented all the methods necessary to fetch, adding and deleting product tags.

## Changes
- Implemented `ProductTagListMapper` + tests. This mapper is able to handle also the tags under a batch update response (which contains keys like `create` or `delete`).
- `ProductTag` model now contains the `siteID` field.
- Implemented `ProductTagsRemote` which implements all the methods to load product tags, create product tags and remove product tags. Added also the tests.
- Added some JSON responses for product tags.
- New Core Data model 29: added `siteID` attribute to `ProductTag` entity, updated `ProductTag`'s  `product` relationship to `products`, updated `Product`'s `tags` relationship with `nullify` as delete rule, implemented `WooCommerceModelV28toV29.xcmappingmodel` to remove product tags without `siteID`.
- Added migration test model from 28 to 29. 
- Implemented some utility methods in `StorageType+Deletions` for deleting product tags.
- Implemented some utility methods in `StorageType+Extensions` for loading product tags.
- Implemented `ProductTagAction` + `ProductTagStore`, which allows also to retrieve all the product tags for the specified site id. Added the tests for `ProductTagStore`.

## Testing
- Look at the code
- CI
Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

jaclync commented on Yosemite/YosemiteTests/Stores/ProductStoreTests.swift:
```diff
@@ -550,7 +550,7 @@ final class ProductStoreTests: XCTestCase {
         productStore.upsertStoredProduct(readOnlyProduct: sampleProduct(), in: viewStorage)
         XCTAssertEqual(viewStorage.countObjects(ofType: Storage.Product.self), 1)
-        XCTAssertEqual(viewStorage.countObjects(ofType: Storage.ProductTag.self), 9)
+        XCTAssertEqual(viewStorage.countObjects(ofType: Storage.ProductTag.self, matching: NSPredicate(format: "siteID == %lld", sampleSiteID)), 9)
```
does the test insert any tags not for the product's site? just wondering because all the other relationships don't require this site predicate

jaclync replied:
this or a separate test case could also be a good place to assert the order of the product tags when upserting to Storage layer

pmusolino replied:
> does the test insert any tags not for the product's site? just wondering because all the other relationships don't require this site predicate

Since `ProductTag` can exist also if they are not related to a specific product, this means that if we previously add some tags, the count will be wrong.

> this or a separate test case could also be a good place to assert the order of the product tags when upserting to Storage layer

I'm implementing a separate test for this case. 👍

pmusolino replied:
Yes, we did the reset at the beginning of every test case, but in these test specific test cases, we want to demonstrate that an existing product, if edited, change their params, including tags. For example for one test case `testUpdateStoredProductEffectivelyUpdatesCorrectSitesProduct()`, we want to store a sample product, make sure that we stored 9 tags for it. We also add a new product with a different site ID, and make sure that the number of tags was not summed with the first one. In the end, we mutate the initial sample product, removing some tags, and make sure that at the ends the tags related to this product are 5. In this case, if we don't use a predicate we need to compare the value with the sum of all the tags added during this test, which is a little bit strange.

jaclync commented on Storage/Storage/Tools/StorageType+Deletions.swift:
```diff
@@ -54,4 +54,29 @@ public extension StorageType {
+    /// Deletes all of the stored Product Tags that don't have an active product relationship
+    ///
+    func deleteUnusedProductTags(siteID: Int64) {
+        let tagsWithNoAssociatedProducts = loadProductTags(siteID: siteID).filter { tag in
```
nit: probably can make this one line like `loadProductTags(siteID: siteID).filter { $0.products?.isEmpty == false }` ?

pmusolino replied:
I think you indented `== true` 😃  edited 👍

jaclync replied:
actually, if `$0.products` is nil, this is going to be false. maybe `isNilOrEmpty`?

pmusolino replied:
Fixed here 9c5c6a5139bcc56590ef8533231b9680489bf651

jaclync commented on Storage/Storage/Tools/StorageType+Deletions.swift:
```diff
@@ -54,4 +54,29 @@ public extension StorageType {
+    /// Deletes all the stored Product Tags with the specified IDs
+    ///
+    func deleteProductTags(siteID: Int64, ids: [Int64]) {
+        let tagsWithSpecifiedIDs = loadProductTags(siteID: siteID).filter { tag in
```
nit: probably can make this one line like `loadProductTags(siteID: siteID).filter { ids.contains($0.tagID) }`?

pmusolino commented on Networking/Networking/Mapper/ProductTagListMapper.swift:
```diff
@@ -0,0 +1,64 @@
+
+        return decodedResponse?.tags ??
+            decodedResponseBatchUpdatedTags?.createdTags ??
+            decodedResponseBatchUpdatedTags?.deletedTags ?? []
```
> it looks like the major difference between `ProductTagListEnvelope` and `ProductTagListBatchUpdateEnvelope` is the key that maps to the tags data

In reality, this is not the only difference, because `create` and `delete` are nested under `data`.

`ProductTagListEnvelope` is used to parse a JSON like that, where `data` is the only key (like all the others endpoints):
```
{
    "data": [
      {
        "id": 34,
        "name": "Leather Shoes",
        "slug": "leather-shoes",
        "description": "",
        "count": 0,
        "_links": {
          "self": [
            {
              "href": "https://example.com/wp-json/wc/v3/products/tags/34"
            }
          ],
          "collection": [
            {
              "href": "https://example.com/wp-json/wc/v3/products/tags"
            }
          ]
        }
      }
}
```

`ProductTagListBatchUpdateEnvelope` is a little bit different, and it's the first time that we need to parse a case like that. Basically, the result are nested keys under `data`, so passing an enum of `CodingKeys` it's not really helpful.
This is a JSON example:
```
{
    "data": {
      "create": [
        {
          "id": 36,
          "name": "Round toe",
          "slug": "round-toe",
          "description": "",
          "count": 0,
          "_links": {
            "self": [
              {
                "href": "https://example.com/wp-json/wc/v3/products/tags/36"
              }
            ],
            "collection": [
              {
                "href": "https://example.com/wp-json/wc/v3/products/tags"
              }
            ]
          }
        }
      ]
    }
}

```

jaclync replied:
thanks for the JSON examples, they were helpful for understanding the response structure!

I just felt that it could be more efficient and explicit if we don't have try decoding with both types of envelope for all types of responses, by passing in an enum for the type of response the caller expects. something like:

```swift
enum ProductTagListMapper.ResponseType {
  case load
  case create
  case delete
}
```

so that `ProductTagListMapper` can be initialized as `ProductTagListMapper(siteID:responseType:)` in `ProductTagsRemote`.

then we have three types of envelope: the existing `ProductTagListEnvelope` for `load` case, and splitting `ProductTagListBatchUpdateEnvelope` into two - `ProductTagListBatchCreateEnvelope` for `create` case and `ProductTagListBatchDeleteEnvelope` for `delete` case.

in this `func map(response: Data) throws -> [ProductTag]`, we can now return the response based on the case:

```swift
switch responseType {
  case .load:
    return try decoder.decode(ProductTagListEnvelope.self, from: response).tags
  case .create:
    return try decoder.decode(ProductTagListBatchCreateEnvelope.self, from: response).tags
  case .delete:
    return try decoder.decode(ProductTagListBatchDeleteEnvelope.self, from: response).tags
}
```

what do you think?

pmusolino replied:
Sounds good to me! Updated here 2f8851766fd2539fbf9ed23c0e23cd0ddfb3479f and here 9711995cb027cea06724df3df32ae5763839790a

pmusolino commented on Yosemite/Yosemite/Actions/ProductTagAction.swift:
```diff
@@ -0,0 +1,33 @@
+    /// Synchronizes all ProductTags matching the specified criteria.
+    /// `onCompletion` will be invoked when the sync operation finishes.
+    ///
+    case synchronizeAllProductTags(siteID: Int64, fromPageNumber: Int, onCompletion: (ProductTagActionError?) -> Void)
```
I'd say that this method is not born for being used with the `SyncingCoordinator`, since the implementation works to avoid it. We want to fetch all the tags, hiding the pagination. `fromPageNumber` is here because I use it to recover the pagination from a page in the case at a certain point the fetching fails. Using this page here, it's useful to avoid errors.

> also, would be nice if we can use Result<[ProductTag], Error> in the completion block!

This is a particular case, where explicitly I didn't use `Result<>`, because here I just need a `ProductTagActionError` as completion in case the synchronization failed.  Using `ProductTagActionError`, I can recover the synchronization from a specified `pageNumber` (as mentioned above).

jaclync replied:
this makes sense now since we aren't using pagination for this action 👍

pmusolino commented on Yosemite/Yosemite/Stores/ProductTagStore.swift:
```diff
@@ -0,0 +1,195 @@
+                    return
+                }
+
+                // Request the next page recursively
```
Yes it is intentional. The main reason is that while the user is typing we need to show all the existing tags with that name, and we can do that correctly only fetching all the existing tags.

jaclync replied:
ah I see, I just checked WPiOS app and it looks like it fetches all the tags at once as well.

in this case, `pageNumber` seems like an implementation detail and we probably always want to sync from the first page? would it be more clear if `pageNumber` is hidden from outside of this function, removing it from `ProductTagAction` and adding a default value to the first page number in this function? otherwise, it feels like this action is meant to work with pagination :)

pmusolino replied:
You are right, and your solution seems a really good idea! Done here 👍 6df8efcf98ea8ed45690544cde2bceaf26072939

pmusolino commented on Yosemite/Yosemite/Stores/ProductStore.swift:
```diff
@@ -445,28 +445,21 @@ private extension ProductStore {
-                                                               tagID: readOnlyTag.tagID) {
+            if let existingStorageTag = storage.loadProductTag(siteID: siteID, tagID: readOnlyTag.tagID) {
                 existingStorageTag.update(with: readOnlyTag)
+                storageProduct.addToTags(existingStorageTag)
```
Good point. I converted tags to `NSOrderedSet`, and made all the relative changes. I also added a test for it. Thank you 💯 for noticing it.

pmusolino commented:
Thank you so much @jaclync for your review!

> Took some time to review every file 😅 it feels like this PR could be separated into Storage layer changes first, then Networking and Yosemite layer changes next time? usually the Storage layer review is pretty quick 🙂

You are totally right @jaclync, unfortunately, I started to touch the Storage layer thinking that there were small changes, but I was wrong. It was an evaluation error 😅

> Since GitHub doesn't display the diffs for the v28 to v29 mapping model, just commenting here: it probably has no effect, but maybe we can remove the entry of ProductTag in the mapping model?

If I understood what you mean, I think I did exactly this. I removed the `ProductTag` in the mapping model (`WooCommerceModelV28toV29.xcmappingmodel`), in this way I removed all the old `ProductTag` models without the `siteID` field 👍 


I made all the changes following your suggestions. Can you take a second look at it please? 🙇‍♂️

pmusolino commented:
> ah sorry I wasn't clear before, it looks like there is still a ProductTag entry in the mapping model XML file. I think it's a bug in the Xcode editor that it doesn't delete it in the XML?

Ohh you are right, there is an entry about `ProductTag` 

```
<object type="XDDEVENTITYMAPPING" id="z215">
        <attribute name="sourcename" type="string">ProductTag</attribute>
        <attribute name="mappingtypename" type="string">Undefined</attribute>
        <attribute name="mappingnumber" type="int16">1</attribute>
        <attribute name="destinationname" type="string">ProductTag</attribute>
        <attribute name="autogenerateexpression" type="bool">1</attribute>
        <relationship name="mappingmodel" type="1/1" destination="XDDEVMAPPINGMODEL"></relationship>
        <relationship name="attributemappings" type="0/0" destination="XDDEVATTRIBUTEMAPPING" idrefs="z549 z364 z370 z433"></relationship>
        <relationship name="relationshipmappings" type="0/0" destination="XDDEVRELATIONSHIPMAPPING" idrefs="z165"></relationship>
    </object>
```

But I'm not sure if it's correct or not to remove it since the mapping model works as expected, and it's not showed in Xcode. Consider that, after I changed tags to be of type `NSOrderedSet` the xcmappingmodel no longer worked (the test failed), and I was forced to regenerate it. So I don't think that's entirely superfluous.

pmusolino commented:
I'm going to merge it right now 👍
